src/memory/memory_reservation.cpp

```diff
@@ -125,272 +189,398 @@ std::unique_ptr<memory_reservation_manager> memory_reservation_manager::_instanc
-    : tier(t), device_id(dev_id), memory_limit(mem_limit), allocators(std::move(allocs)) {
-    if (allocators.empty()) {
-        throw std::invalid_argument("At least one allocator must be provided");
-    }
+  Tier t,
+  size_t dev_id,
+  size_t mem_limit,
+  std::vector<std::unique_ptr<rmm::mr::device_memory_resource>> allocs)
+    : tier(t)
+    , device_id(dev_id)
+    , memory_limit(mem_limit)
+    , allocators(std::move(allocs))
+{
+  if (allocators.empty())
+  {
+    throw std::invalid_argument("At least one allocator must be provided");
+  }
 }
 
-void memory_reservation_manager::initialize(std::vector<memory_space_config> configs) {
-    // Test hook: if a test called reset_for_testing(), allow reinitialization bypassing call_once
-    if (_allow_reinitialize_for_tests) {
-        _allow_reinitialize_for_tests = false;
-        _instance = std::unique_ptr<memory_reservation_manager>(new memory_reservation_manager(std::move(configs)));
-        return;
-    }
-    std::call_once(_initialized, [configs = std::move(configs)]() mutable {
-        _instance = std::unique_ptr<memory_reservation_manager>(new memory_reservation_manager(std::move(configs)));
-    });
+void memory_reservation_manager::initialize(std::vector<memory_space_config> configs)
+{
+  // Test hook: if a test called reset_for_testing(), allow reinitialization bypassing call_once
+  if (_allow_reinitialize_for_tests)
+  {
+    _allow_reinitialize_for_tests = false;
+    _instance                     = std::unique_ptr<memory_reservation_manager>(
+      new memory_reservation_manager(std::move(configs)));
+    return;
+  }
+  std::call_once(_initialized, [configs = std::move(configs)]() mutable {
+    _instance = std::unique_ptr<memory_reservation_manager>(
+      new memory_reservation_manager(std::move(configs)));
+  });
 }
 
-void memory_reservation_manager::reset_for_testing() {
-    // Not thread-safe; intended for unit tests only
-    _instance.reset();
-    _allow_reinitialize_for_tests = true;
-}
-
-memory_reservation_manager& memory_reservation_manager::get_instance() {
-    if (!_instance) {
-        throw std::runtime_error("memory_reservation_manager not initialized. Call initialize() first.");
+void memory_reservation_manager::topo_aware_initialize()
+{
+  rapidsmpf::experimental::TopologyDiscovery discovery;
+  if (!discovery.discover())
+  {
+    throw std::runtime_error("Failed to discover system topology");
+  };
+
+  // Get the discovered topology
+
+  // TODO: what is the GPU allocation policy? Single process managing many devices?
+  // One-device-per-process?
+  // TODO: How would we handle CUDA_VISIBLE_DEVICES env var?
+  const auto& topology = discovery.get_topology();
+
+  // Save and restore CUDA device context using RAII
+  scoped_device_context device_guard;
+
+  // Create a vector of memory_space_config for all GPUs
+  std::vector<memory_space_config> configs;
+  configs.reserve(topology.num_gpus);
+
```

Comment:
We also need to make memory spaces for the different numa nodes. Given that all memory spaces are connected to numa nodes we should probably add another value to memory space which is the numa ID of the node. The memory spaces for host should use https://github.com/sirius-db/sirius/blob/dev/src/include/memory/fixed_size_host_memory_resource.hpp as their allocator